include/swift/DependencyScan/DependencyScanImpl.h

```diff
@@ -0,0 +1,126 @@
+//
+//===----------------------------------------------------------------------===//
+#ifndef SWIFT_C_DEPENDENCY_SCAN_IMPL_H
+#define SWIFT_C_DEPENDENCY_SCAN_IMPL_H
+
+#include "swift-c/DependencyScan/DependencyScan.h"
+#include "swift/DependencyScan/DependencyScanningTool.h"
+
+DEPSCAN_BEGIN_DECLS
+
+/// Swift modules to be built from a module interface, may have a bridging
+/// header.
+typedef struct {
+  /// The module interface from which this module was built, if any.
+  ds_string_t module_interface_path;
+
+  /// The paths of potentially ready-to-use compiled modules for the interface.
+  ds_string_set_t *compiled_module_candidates;
+
+  /// The bridging header, if any.
+  ds_string_t bridging_header_path;
+
+  /// The source files referenced by the bridging header.
+  ds_string_set_t *bridging_source_files;
+
+  /// (Clang) modules on which the bridging header depends.
+  ds_string_set_t *bridging_module_dependencies;
+
+  /// Options to the compile command required to build this module interface
+  ds_string_set_t *command_line;
+
+  /// To build a PCM to be used by this Swift module, we need to append these
+  /// arguments to the generic PCM build arguments reported from the dependency
+  /// graph.
+  ds_string_set_t *extra_pcm_args;
+
+  /// The hash value that will be used for the generated module
+  ds_string_t context_hash;
+
+  /// A flag to indicate whether or not this module is a framework.
+  bool is_framework;
+} ds_swift_textual_details_t;
+
+/// Swift modules with only a binary module file.
+typedef struct {
+  /// The path to the pre-compiled binary module
+  ds_string_t compiled_module_path;
+
+  /// The path to the .swiftModuleDoc file.
+  ds_string_t module_doc_path;
+
+  /// The path to the .swiftSourceInfo file.
+  ds_string_t module_source_info_path;
+} ds_swift_binary_details_t;
+
+/// Swift placeholder modules carry additional details that specify their
+/// module doc path and source info paths.
+typedef struct {
+  /// The path to the pre-compiled binary module
+  ds_string_t compiled_module_path;
+
+  /// The path to the .swiftModuleDoc file.
+  ds_string_t module_doc_path;
+
+  /// The path to the .swiftSourceInfo file.
+  ds_string_t module_source_info_path;
+} ds_swift_placeholder_details_t;
+
+/// Clang modules are built from a module map file.
+typedef struct {
+  /// The path to the module map used to build this module.
+  ds_string_t module_map_path;
+
+  /// clang-generated context hash
+  ds_string_t context_hash;
+
+  /// Options to the compile command required to build this clang modulemap
+  ds_string_set_t *command_line;
+} ds_clang_details_t;
+
+typedef struct {
+  ds_dependency_info_kind_t kind;
+  union {
+    ds_swift_textual_details_t swift_textual_details;
+    ds_swift_binary_details_t swift_binary_details;
+    ds_swift_placeholder_details_t swift_placeholder_details;
+    ds_clang_details_t clang_details;
+  };
+} ds_impl_module_details_t;
```

Comment:
Since you already have the underlying types of the exported opaque types as C structs (instead of C++) you could make things a bit more convenient on the C++ side by exporting the types as pointers to structs, like this:
```
// in the exported C header
typedef struct ds_module_details_s *ds_module_details_t;

[...]

// in the implementation
struct ds_module_details_s {
  ds_dependency_info_kind_t kind;
  union {..}
};
```
The benefit of this is you have C++ type safety and you don't need a bunch of 'wrap/unwrap' functions to call.
Note that this doesn't make a difference for the Swift side, the pointers will still be opaque when imported in Swift.

Reply:
This is indeed a lot cleaner and the safety is nice to have. Thanks for pointing out this technique.